File: leetcode/two-pointers/Container With Most Water.py
# https://leetcode.com/problems/container-with-most-water/description/
# ⭐️

class Solution:
    def maxArea(self, height: List[int]) -> int:
        l, r = 0, len(height) - 1
        res = 0
        while l < r:
            res = max(res, min(height[l], height[r]) * (r - l)) # 가로조심 연산순서바뀐다
            if height[l] <= height[r]:
                l += 1
            elif height[r] < height[l]:
                r -= 1
            
        return res

File: neetcode/go/Two Pointers/Container With Most Water.go
package twopointers

func maxArea(heights []int) int {
	l, r := 0, len(heights)-1
	res := 0

	for l < r {
		area := min(heights[l], heights[r]) * (r - l)
		if area > res {
			res = area
		}
		if heights[l] <= heights[r] {
			l++
		} else {
			r--
		}
	}
	return res
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
